File: src/main/java/no/novari/flyt/kafka/instanceflow/consuming/InstanceFlowErrorHandlerConfigurationMapper.java
package no.novari.flyt.kafka.instanceflow.consuming;

import lombok.extern.slf4j.Slf4j;
import no.novari.flyt.kafka.instanceflow.headers.CouldNotReadInstanceFlowHeadersException;
import no.novari.flyt.kafka.instanceflow.headers.NoInstanceFlowHeadersException;
import no.novari.kafka.consuming.ErrorHandlerConfiguration;
import org.springframework.stereotype.Service;

@Slf4j
@Service
public class InstanceFlowErrorHandlerConfigurationMapper {

    private final InstanceFlowConsumerRecordMapper instanceFlowConsumerRecordMapper;

    public InstanceFlowErrorHandlerConfigurationMapper(
            InstanceFlowConsumerRecordMapper instanceFlowConsumerRecordMapper
    ) {
        this.instanceFlowConsumerRecordMapper = instanceFlowConsumerRecordMapper;
    }

    public <VALUE> ErrorHandlerConfiguration<VALUE> toErrorHandlerConfiguration(
            InstanceFlowErrorHandlerConfiguration<VALUE> instanceFlowErrorHandlerConfiguration
    ) {
        ErrorHandlerConfiguration.ErrorHandlerConfigurationBuilder<VALUE>
                errorHandlerConfigurationBuilder = ErrorHandlerConfiguration.builder();

        instanceFlowErrorHandlerConfiguration
                .getBackOffFunction()
                .ifPresent(backOffFunction ->
                        errorHandlerConfigurationBuilder.backOffFunction(
                                (consumerRecord, exception) ->
                                        backOffFunction.apply(
                                                instanceFlowConsumerRecordMapper.toInstanceFlowConsumerRecord(
                                                        consumerRecord),
                                                exception
                                        )
                        )
                );

        instanceFlowErrorHandlerConfiguration
                .getDefaultBackoff()
                .ifPresent(errorHandlerConfigurationBuilder::defaultBackoff);

        instanceFlowErrorHandlerConfiguration
                .getCustomRecoverer()
                .ifPresent(recoverer ->
                        errorHandlerConfigurationBuilder.customRecoverer(
                                (consumerRecord, consumer, exception) -> {
                                    try {
                                        recoverer.recover(
                                                instanceFlowConsumerRecordMapper
                                                        .toInstanceFlowConsumerRecord(consumerRecord),
                                                exception
                                        );
                                    } catch (NoInstanceFlowHeadersException |
                                             CouldNotReadInstanceFlowHeadersException e) {
                                        log.warn("Calling recoverer without instance flow headers", e);
                                        recoverer.recoverForMissingInstanceFlowHeaders(
                                                consumerRecord,
                                                exception
                                        );
                                    }
                                }
                        )
                );

        return errorHandlerConfigurationBuilder
                .classificationType(instanceFlowErrorHandlerConfiguration.getClassificationType())
                .classificationExceptions(instanceFlowErrorHandlerConfiguration.getClassificationExceptions())
                .skipRecordOnRecoveryFailure(instanceFlowErrorHandlerConfiguration.isSkipRecordOnRecoveryFailure())
                .restartRetryOnExceptionChange(instanceFlowErrorHandlerConfiguration.isRestartRetryOnExceptionChange())
                .restartRetryOnRecoveryFailure(instanceFlowErrorHandlerConfiguration.isRestartRetryOnRecoveryFailure())
                .build();
    }

}

File: src/main/kotlin/no/novari/flyt/kafka/instanceflow/consuming/InstanceFlowErrorHandlerConfigurationMapper.kt
package no.novari.flyt.kafka.instanceflow.consuming

import no.novari.flyt.kafka.instanceflow.headers.CouldNotReadInstanceFlowHeadersException
import no.novari.flyt.kafka.instanceflow.headers.NoInstanceFlowHeadersException
import no.novari.kafka.consuming.ErrorHandlerConfiguration
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.util.Optional

@Service
class InstanceFlowErrorHandlerConfigurationMapper(
    private val instanceFlowConsumerRecordMapper: InstanceFlowConsumerRecordMapper,
) {
    fun <VALUE> toErrorHandlerConfiguration(
        instanceFlowErrorHandlerConfiguration: InstanceFlowErrorHandlerConfiguration<VALUE>,
    ): ErrorHandlerConfiguration<VALUE> {
        val builder = ErrorHandlerConfiguration.builder<VALUE>()

        instanceFlowErrorHandlerConfiguration.backOffFunction?.let { backOffFunction ->
            builder.backOffFunction { consumerRecord, exception ->
                Optional.ofNullable(
                    backOffFunction.apply(
                        instanceFlowConsumerRecordMapper.toInstanceFlowConsumerRecord(consumerRecord),
                        exception,
                    ),
                )
            }
        }

        instanceFlowErrorHandlerConfiguration.defaultBackoff?.let(builder::defaultBackoff)

        instanceFlowErrorHandlerConfiguration.customRecoverer?.let { recoverer ->
            builder.customRecoverer {
                consumerRecord,
                _,
                exception,
                ->
                try {
                    recoverer.recover(
                        instanceFlowConsumerRecordMapper.toInstanceFlowConsumerRecord(consumerRecord),
                        exception,
                    )
                } catch (e: NoInstanceFlowHeadersException) {
                    log.warn("Calling recoverer without instance flow headers", e)
                    recoverer.recoverForMissingInstanceFlowHeaders(consumerRecord, exception)
                } catch (e: CouldNotReadInstanceFlowHeadersException) {
                    log.warn("Calling recoverer without instance flow headers", e)
                    recoverer.recoverForMissingInstanceFlowHeaders(consumerRecord, exception)
                }
            }
        }

        return builder
            .classificationType(instanceFlowErrorHandlerConfiguration.classificationType)
            .classificationExceptions(instanceFlowErrorHandlerConfiguration.classificationExceptions)
            .skipRecordOnRecoveryFailure(instanceFlowErrorHandlerConfiguration.skipRecordOnRecoveryFailure)
            .restartRetryOnExceptionChange(instanceFlowErrorHandlerConfiguration.restartRetryOnExceptionChange)
            .restartRetryOnRecoveryFailure(instanceFlowErrorHandlerConfiguration.restartRetryOnRecoveryFailure)
            .build()
    }

    companion object {
        private val log = LoggerFactory.getLogger(InstanceFlowErrorHandlerConfigurationMapper::class.java)
    }
}
